File: app/src/main/java/com/wxm/camerajob/ui/Base/SendEmailHelper.java
package com.wxm.camerajob.ui.Base;

import android.util.Log;

import java.util.*;

import javax.activation.CommandMap;
import javax.activation.MailcapCommandMap;
import javax.mail.*;
import javax.mail.internet.*;

/**
 * helper for send email
 * Created by 123 on 2016/11/10.
 */
public class SendEmailHelper extends javax.mail.Authenticator {
    private final static String    TAG = "SendEmailHelper";
    private SendEmailPara   mSEPara;

    private MimeMessage     mMMMsg;
    private Properties      mPPprops;

    public SendEmailHelper()    {
    }

    @Override
    public PasswordAuthentication getPasswordAuthentication() {
        return new PasswordAuthentication(mSEPara.mSendUsr, mSEPara.mSendPWD);
    }

    public void sendEmail(SendEmailPara sp) {
        mSEPara = sp;
        Thread thread = new Thread(() -> {
            try {
                if(init_context()) {
                    if(send_out())
                        mSEPara.mIFOnResult.onSendSuccess();
                    else
                        mSEPara.mIFOnResult.onSendFailure();
                } else {
                    mSEPara.mIFOnResult.onSendFailure();
                }
            } catch (Exception e) {
                Log.e(TAG, e.getMessage());
            }
        });
        thread.start();
    }

    private boolean send_out() {
        try {
            Transport.send(mMMMsg);
        } catch (Exception e) {
            e.printStackTrace();
            return false;
        }

        return true;
    }

    private boolean init_context()   {
        // for system property
        mPPprops = System.getProperties();
        mPPprops.put("mail.smtp.host", mSEPara.mSendServerHost);
        mPPprops.put("mail.smtp.auth", "true");
        mPPprops.put("mail.debug", "true");

        mPPprops.put("mail.smtp.socketFactory.port", "465");
        mPPprops.put("mail.smtp.port", "465");
        mPPprops.put("mail.smtp.socketFactory.class", "javax.net.ssl.SSLSocketFactory");
        mPPprops.put("mail.smtp.socketFactory.fallback", "false");


        // for holder
        Session mSNsession = Session.getDefaultInstance(mPPprops, this);
        mMMMsg = new MimeMessage(mSNsession);
        Multipart mMPparts = new MimeMultipart();

        // There is something wrong with MailCap, javamail can not find a handler for the multipart/mixed part, so this bit needs to be added.
        MailcapCommandMap mc = (MailcapCommandMap) CommandMap.getDefaultCommandMap();
        mc.addMailcap("text/html;; x-java-content-handler=com.sun.mail.handlers.text_html");
        mc.addMailcap("text/xml;; x-java-content-handler=com.sun.mail.handlers.text_xml");
        mc.addMailcap("text/plain;; x-java-content-handler=com.sun.mail.handlers.text_plain");
        mc.addMailcap("multipart/*;; x-java-content-handler=com.sun.mail.handlers.multipart_mixed");
        mc.addMailcap("message/rfc822;; x-java-content-handler=com.sun.mail.handlers.message_rfc822");
        CommandMap.setDefaultCommandMap(mc);

        // for email info
        try {
            // setup message body
            BodyPart messageBodyPart = new MimeBodyPart();
            messageBodyPart.setText(mSEPara.mEmailBody);
            mMPparts.addBodyPart(messageBodyPart);

            mMMMsg.setSubject(mSEPara.mEmailTitle);
            mMMMsg.setFrom(new InternetAddress(mSEPara.mSendUsr));
            mMMMsg.setRecipients(Message.RecipientType.TO,
                    InternetAddress.parse(mSEPara.mRecvUsr));
            mMMMsg.setSentDate(new Date());
            mMMMsg.setContent(mMPparts);
            mMMMsg.saveChanges();
        } catch (MessagingException e) {
            e.printStackTrace();
            return false;
        }

        return true;
    }
}

File: app/src/main/java/com/wxm/camerajob/ui/base/SendEmailHelper.kt
package com.wxm.camerajob.ui.base

import wxm.androidutil.log.TagLog
import java.util.*
import javax.activation.CommandMap
import javax.activation.MailcapCommandMap
import javax.mail.*
import javax.mail.internet.InternetAddress
import javax.mail.internet.MimeBodyPart
import javax.mail.internet.MimeMessage
import javax.mail.internet.MimeMultipart

/**
 * helper for send email
 * Created by 123 on 2016/11/10.
 */
class SendEmailHelper : javax.mail.Authenticator() {
    private var mSEPara: SendEmailPara? = null
    private var mMMMsg: MimeMessage? = null

    public override fun getPasswordAuthentication(): PasswordAuthentication {
        return PasswordAuthentication(mSEPara!!.mSendUsr, mSEPara!!.mSendPWD)
    }

    fun sendEmail(sp: SendEmailPara) {
        mSEPara = sp
        val thread = Thread {
            try {
                if (initContext()) {
                    if (sendOut())
                        mSEPara!!.mIFOnResult!!.onSendSuccess()
                    else
                        mSEPara!!.mIFOnResult!!.onSendFailure()
                } else {
                    mSEPara!!.mIFOnResult!!.onSendFailure()
                }
            } catch (e: Exception) {
                TagLog.e("send email failure", e)
            }
        }
        thread.start()
    }

    private fun sendOut(): Boolean {
        try {
            Transport.send(mMMMsg!!)
        } catch (e: Exception) {
            e.printStackTrace()
            return false
        }

        return true
    }

    private fun initContext(): Boolean {
        // for system property
        val mAttr = System.getProperties().let {
            it[MAIL_HOST] = mSEPara!!.mSendServerHost
            it[MAIL_AUTH] = "true"
            it[MAIL_DEBUG] = "true"
            it[MAIL_PORT] = "465"

            it[MAIL_SOCKETFACTORY_PORT] = "465"
            it[MAIL_SOCKETFACTORY_CLASS] = "javax.net.ssl.SSLSocketFactory"
            it[MAIL_SOCKETFACTORY_FALLBACK] = "false"

            it
        }

        // for holder
        val mSNsession = Session.getDefaultInstance(mAttr, this)
        mMMMsg = MimeMessage(mSNsession)

        // There is something wrong with MailCap, javamail can not find a handler for the multipart/mixed part, so this bit needs to be added.
        (CommandMap.getDefaultCommandMap() as MailcapCommandMap).apply {
            addMailcap("text/html;; x-java-content-handler=com.sun.mail.handlers.text_html")
            addMailcap("text/xml;; x-java-content-handler=com.sun.mail.handlers.text_xml")
            addMailcap("text/plain;; x-java-content-handler=com.sun.mail.handlers.text_plain")
            addMailcap("multipart/*;; x-java-content-handler=com.sun.mail.handlers.multipart_mixed")
            addMailcap("message/rfc822;; x-java-content-handler=com.sun.mail.handlers.message_rfc822")
            CommandMap.setDefaultCommandMap(this)
        }

        // for email info
        try {
            // setup message body
            mMMMsg!!.apply {
                subject = mSEPara!!.mEmailTitle
                setFrom(InternetAddress(mSEPara!!.mSendUsr))
                setRecipients(Message.RecipientType.TO, InternetAddress.parse(mSEPara!!.mRecvUsr))
                sentDate = Date()
                setContent(MimeMultipart().apply {
                    addBodyPart(MimeBodyPart().apply { setText(mSEPara!!.mEmailBody) })
                })
                saveChanges()
            }
        } catch (e: MessagingException) {
            e.printStackTrace()
            return false
        }

        return true
    }

    companion object {
        private const val MAIL_HOST = "mail.smtp.host"
        private const val MAIL_AUTH = "mail.smtp.auth"
        private const val MAIL_DEBUG = "mail.debug"

        private const val MAIL_PORT = "mail.smtp.port"
        private const val MAIL_SOCKETFACTORY_PORT = "mail.smtp.socketFactory.port"
        private const val MAIL_SOCKETFACTORY_CLASS = "mail.smtp.socketFactory.port"
        private const val MAIL_SOCKETFACTORY_FALLBACK = "mail.smtp.socketFactory.fallback"
    }
}


